packages/flutter/lib/src/material/reorderable_list.dart

```diff
@@ -0,0 +1,517 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math';
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter/rendering.dart';
+
+import 'material.dart';
+
+/// The callback used by [ReorderableListView] to move an item to a new
+/// position in a list.
+/// 
+/// Implementations should remove the corresponding list item at [oldIndex]
+/// and reinsert it at [newIndex].
+/// 
+/// Note that if [oldIndex] is before [newIndex], removing the item at [oldIndex]
+/// from the list will reduce the list's length by one. Implementations used
+/// by [ReorderableListView] will need to account for this when inserting before
+/// [newIndex].
+/// 
+/// Example implementation:
+/// 
+/// ```dart
+/// final List<MyDataObject> backingList = <MyDataObject>[/* ... */];
+/// 
+/// void onReorder(int oldIndex, int newIndex) {
+///   if (oldIndex < newIndex) {
```

Comment:
This is an O(2*N) operation with an array. Could a more efficient example be provided that swaps the two items?

```
onReorder: (fromIndex, toIndex) {
  setState(() {
    _swapInPlace(_myReorderableList, fromIndex, toIndex - 1);
  });
}

void _swapInPlace(List list, int index1, int index2) {
  final swap = list[index1];
  list[index1] = list[index2];
  list[index2] = list[index1];
}
```

Maybe even consider changing the API so you don't have to subtract 1 from the to index?

Reply:
Please ignore it's late I need to go home :)